### PR TITLE
Fixes #95: Create listDevices function and fail with multiple devices

### LINKED ISSUE #95: Fail if more than one device is connected
As long as we don’t support #1, yet, let’s fails if there is more than one device. This includes devices of more than one platform, see:

              Might be a nice UX feature to fail in `ensureDevice()` if there is more than one device detected by adb and/or Frida.

_Originally posted by @baltpeter in https://github.com/tweaselORG/appstraction/issues/94#issuecomment-1593049800_

### PR DESCRIPTION
**Pull request by zner0L**

### CODE REVIEW

File: src/util.ts [hunk 1]
```diff
@@ -324,3 +326,86 @@ export const parsePemCertificateFromFile = async (path: string) => {
...
+    results.push(
+        ...androidDevices.map((device) => ({
+            id: device.id,
+            name: device.name,
+            platform: 'android' as const,
+        }))
+    );
+
+    const iosDevices = (await python('pymobiledevice3', ['usbmux', 'list', '--no-color'], { reject: false }).then(
+        ({ stdout, exitCode }) => (exitCode !== 0 ? JSON.parse(stdout) : [])
```
Comment (baltpeter) [inline src/util.ts:403]: The condition is inverted. This currently never returns any iOS devices.

File: src/android.ts [hunk 2]
```diff
@@ -295,6 +296,22 @@ export const androidApi = <RunTarget extends SupportedRunTarget<'android'>>(
...
+
+        const availableDevices = await listDevices({ frida: options.capabilities.includes('frida') });
+
+        if (availableDevices.length > 1)
+            throw new Error('You have multiple devices connected. Please disconnect all but one.');
+        else if (availableDevices.length === 0)
+            throw new Error(
+                options.runTarget === 'device' ? 'You need to connect your device.' : 'You need to start the emulator.'
+            );
+        else if (availableDevices.filter((device) => device.platform === 'ios').length === 0)
```
Comment (baltpeter) [inline src/android.ts:308]:
```suggestion
        else if (availableDevices.filter((device) => device.platform === 'android').length === 0)
```

File: src/android.ts [hunk 3]
```diff
@@ -295,6 +296,22 @@ export const androidApi = <RunTarget extends SupportedRunTarget<'android'>>(
...
+            throw new Error(
+                options.runTarget === 'device'
+                    ? 'You need to connect an Android device.'
+                    : 'You need to start the emulator.'
+            );
+
         if (
```
Comment (baltpeter) [inline src/android.ts:315]: Shouldn't we have a different error (that mentions that the device needs to be booted) here now?

File: src/util.ts [hunk 4]
```diff
@@ -324,3 +326,86 @@ export const parsePemCertificateFromFile = async (path: string) => {
...
+                            platform: params.platform === 'darwin' ? 'ios' : 'android',
+                        };
+                    })
+                )
+            );
+
+    const venv = getVenv(venvOptions);
+    const python = async (...args: Parameters<Awaited<typeof venv>>) => (await venv)(...args);
+
+    const results = [];
```
Comment (baltpeter) [inline src/util.ts:367]: Why don't you make `aD` and `iD` the correct format already and just return `[...aD, ...iD]` instead of pushing to an array?

**Review by baltpeter — CHANGES_REQUESTED** [review]

File: src/util.ts (hunk 4 above)
Comment (zner0L) [reply src/util.ts:367]: Well, I liked how we could get all that information and wanted to implement a way to make them all available at first, but then I decided against it without changing the code. You are right.

File: src/util.ts [hunk 5]
```diff
@@ -343,3 +345,82 @@ export const escapeCommand = (args: string[]) => {
...
+
+    const androidDevices = await runAndroidDevTool('adb', ['devices', '-l']).then(({ stdout }) => {
+        const lines = stdout.split('\n');
+        if (lines[0]?.includes('List of devices attached'))
+            return lines
+                .slice(1, -1)
+                .map((line) => {
+                    const device = line.split(/\s+/);
+
+                    // Only return devices which are actually connected (possible vaules are 'device', 'offline' and 'no', see https://developer.android.com/tools/adb#devicestatus)
```
Comment (baltpeter) [inline src/util.ts:399]:
```suggestion
                    // Only return devices which are actually connected (possible values are 'device', 'offline' and 'no', see https://developer.android.com/tools/adb#devicestatus)
```

File: src/util.ts [hunk 6]
```diff
@@ -343,3 +345,82 @@ export const escapeCommand = (args: string[]) => {
...
+                    return device[0] && device[1]
+                        ? {
+                              id: device[0],
+                              name: line.match(/model:(.+)? /)?.[1],
+                              platform: 'android' as const,
+                          }
+                        : undefined;
+                })
+                .filter(Boolean) as { id: string; name?: string; platform: 'android' }[];
+        throw new Error('Could not list android devices: unexpected output from adb devices');
```
Comment (baltpeter) [inline src/util.ts:411]: Android throws on unexpected output, iOS ignores it silently. I think ignoring silently is the way to go. If you're only interested in analysing one platform, the setup for the other may be broken and that shouldn't be a problem.

**Review by baltpeter — CHANGES_REQUESTED** [review]

**Review by baltpeter — APPROVED** [review]
Thank you!

**Pull request merged**